File: medium/0235_lowest_common_ancestor_of_a_binary_search_tree/solution.py
from utils.python.nodes import TreeNode


class Solution:
    def lowestCommonAncestor(
        self, root: TreeNode, p: TreeNode, q: TreeNode
    ) -> TreeNode:
        curr = root

        while True:
            if p.val < curr.val and q.val < curr.val:
                curr = curr.left
            elif p.val > curr.val and q.val > curr.val:
                curr = curr.right
            else:
                return curr

        return None

File: medium/0235_lowest_common_ancestor_of_a_binary_search_tree/solution.cpp
#include "../../utils/cpp/nodes.hpp"

class Solution {
public:
  utils::TreeNode *lowestCommonAncestor(utils::TreeNode *root,
                                        utils::TreeNode *p,
                                        utils::TreeNode *q) {
    utils::TreeNode *curr = root;

    while (true) {
      if (p->val < curr->val && q->val < curr->val)
        curr = curr->left;
      else if (p->val > curr->val && q->val > curr->val)
        curr = curr->right;
      else
        return curr;
    }

    return nullptr;
  }
};
